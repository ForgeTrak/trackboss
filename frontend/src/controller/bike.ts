import { generateHeaders } from './utils';
import {
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeRequest,
    PatchBikeResponse,
    PostNewBikeRequest,
    PostNewBikeResponse,
} from '../../../src/typedefs/bike';

export async function createBike(token: string, bikeData: PostNewBikeRequest): Promise<PostNewBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/new`, {
        method: 'POST',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function getBikeList(token: string, membershipID?: number): Promise<GetBikeListResponse> {
    if (membershipID) {
        const idString = membershipID.toString();
        const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list?membershipID=${idString}`, {
            method: 'GET',
            mode: 'cors',
            headers: generateHeaders(token),
        });
        return response.json();
    }
    // else
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list`, {
        method: 'GET',
        mode: 'cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function getBike(token: string, bikeID: number): Promise<GetBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function updateBike(
    token: string,
    bikeID: number,
    bikeData: PatchBikeRequest,
): Promise<PatchBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'PATCH',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function deleteBike(token: string, bikeID: number): Promise<DeleteBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'DELETE',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}
