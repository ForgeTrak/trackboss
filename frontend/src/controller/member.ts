import generateHeaders from "./utils";


export async function createMember(token: string, memberData: object) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/new`, {
        method: 'POST',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(memberData),
    });
    return response.json();
}

export async function getMember(token: string, memberId: number) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/${memberId}`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function getMemberList(token: string, listType?: string) {
    if (listType) {
        const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/list?status=${listType}`, {
            method: 'GET',
            mode: 'no-cors',
            headers: generateHeaders(token),
        });
        return response.json();
    }
    // else
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/list`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function updateMember(token: string, memberID: number, memberData: object) {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/member/${memberID}`, {
        method: 'PATCH',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(memberData),
    });
    return response.json();
}