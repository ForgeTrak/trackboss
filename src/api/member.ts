import { Request, Response, Router } from 'express';

const member = Router();

member.post('/new', (req: Request, res: Response) => {
    res.status(501).send();
});

member.get('/list', (req: Request, res: Response) => {
    res.status(501).send();
});

member.get('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

member.patch('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

member.delete('/:memberID', (req: Request, res: Response) => {
    res.status(501).send();
});

export default member;
