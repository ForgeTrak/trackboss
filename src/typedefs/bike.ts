import { ErrorResponse } from './errorResponse';

export type Bike = {
    bikeId: number,
    year: string,
    make: string,
    model: string,
    membershipAdmin: string
}

export type DeletedBike = {
    bikeId: number
}

export type PostNewBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId: number
}

export type PostNewBikeResponse = Bike | ErrorResponse

export type GetBikeRequest = Record<string, never>

export type GetBikeResponse = Bike | ErrorResponse

export type PatchBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId?: number
}

export type PatchBikeResponse = Bike | ErrorResponse

export type DeleteBikeRequest = Record<string, never>

export type DeleteBikeResponse = DeletedBike | ErrorResponse

export type GetBikeListRequest = Record<string, never>

export type GetBikeListResponse = Bike[] | ErrorResponse
