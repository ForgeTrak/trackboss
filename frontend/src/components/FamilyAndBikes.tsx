import React, { useContext, useEffect, useState } from 'react';
import {
    Heading,
    VStack,
    HStack,
    Divider,
    SimpleGrid,
    Button,
    UnorderedList,
    ListItem,
    useDisclosure,
} from '@chakra-ui/react';
import { GetMemberListResponse, Member } from '../../../src/typedefs/member';
import { Bike, GetBikeListResponse } from '../../../src/typedefs/bike';
import DeleteAlert from './DeleteAlert';
import EditBikesModal from './EditBikeModal';
import AddFamilyModal from './modals/AddFamilyModal';
import AddBikeModal from './AddBikeModal';
import { updateMember } from '../controller/member';
import { UserContext } from '../contexts/UserContext';
import { createBike, deleteBike, getBikeList, updateBike } from '../controller/bike';

interface cardProps {
    memberFamily: GetMemberListResponse,
    memberBikes: GetBikeListResponse,
    admin: boolean
}

export default function GeneralInfo(props: cardProps) {
    const { state } = useContext(UserContext);
    const { onClose: onRemoveFamilyClose, isOpen: isRemoveFamilyOpen, onOpen: onRemoveFamilyOpen } = useDisclosure();
    const { onClose: onAddFamilyClose, isOpen: isAddFamilyOpen, onOpen: onAddFamilyOpen } = useDisclosure();

    const { onClose: onRemoveBikeClose, isOpen: isRemoveBikeOpen, onOpen: onRemoveBikeOpen } = useDisclosure();
    const { onClose: onEditBikeClose, isOpen: isEditBikeOpen, onOpen: onEditBikeOpen } = useDisclosure();
    const { onClose: onAddBikeClose, isOpen: isAddBikeOpen, onOpen: onAddBikeOpen } = useDisclosure();

    const [memberFamily, setMemberFamily] = useState<GetMemberListResponse>([]);
    const [memberBikes, setMemberBikes] = useState<GetBikeListResponse>([]);

    const [memberToRemove, setMemberToRemove] = useState<Member>();
    const [bikeToRemove, setBikeToRemove] = useState<Bike>();
    const [bikeToEdit, setBikeToEdit] = useState<Bike>();

    async function refreshBikeList() {
        const newBikeList = await getBikeList(state.token, state?.user?.membershipId);
        setMemberBikes(newBikeList);
    }

    async function removeFamilyMember() {
        if (memberToRemove !== undefined && state.user !== undefined) {
            await updateMember(
                state.token,
                memberToRemove.memberId,
                { active: false, modifiedBy: state.user.memberId },
            );
        }
    }

    async function removeBike() {
        if (bikeToRemove !== undefined && state.user !== undefined) {
            await deleteBike(state.token, bikeToRemove.bikeId);
        }
        await refreshBikeList();
    }

    async function editBike(editedBike: Bike, bikeYear: string, bikeMake: string, bikeModel: string) {
        await updateBike(
            state.token,
            editedBike.bikeId,
            { year: bikeYear, make: bikeMake, model: bikeModel },
        );
        await refreshBikeList();
    }

    async function addBike(newYear: string, newMake: string, newModel: string) {
        if (state.user !== undefined) {
            await createBike(
                state.token,
                { year: newYear, make: newMake, model: newModel, membershipId: state.user.membershipId },
            );
        }
        await refreshBikeList();
    }

    useEffect(() => {
        async function setMemberData() {
            setMemberFamily(props.memberFamily);
            setMemberBikes(props.memberBikes);
        }
        setMemberData();
    }, [props.memberFamily, props.memberBikes]);

    return (
        <SimpleGrid columns={1}>
            <VStack mt={25} borderRightWidth={0.5} borderRightColor="lightgrey">
                <HStack>
                    <Heading>Family</Heading>
                    {
                        props.admin && (
                            <Button
                                mr={400}
                                textDecoration="underline"
                                color="orange"
                                variant="ghost"
                                size="lg"
                                onClick={
                                    () => {
                                        onAddFamilyOpen();
                                    }
                                }
                            >
                                Add
                            </Button>
                        )
                    }
                </HStack>
                {
                    memberFamily && (
                        <UnorderedList pt={10} spacing={2}>
                            {
                                (memberFamily as Member[]).map((member) => (
                                    <HStack>
                                        <ListItem
                                            fontSize="xl"
                                        >
                                            {` ${member.firstName} ${member.lastName}`}
                                        </ListItem>
                                        {
                                            props.admin && (
                                                <Button
                                                    textDecoration="underline"
                                                    color="red"
                                                    variant="ghost"
                                                    size="sm"
                                                    onClick={
                                                        () => {
                                                            setMemberToRemove(member);
                                                            onRemoveFamilyOpen();
                                                        }
                                                    }
                                                >
                                                    Remove
                                                </Button>
                                            )
                                        }
                                    </HStack>
                                ))
                            }
                        </UnorderedList>
                    )
                }
            </VStack>
            <VStack mt={25}>
                <HStack>
                    <Heading>Bikes</Heading>
                    {
                        props.admin && (
                            <Button
                                textDecoration="underline"
                                color="orange"
                                variant="ghost"
                                size="lg"
                                onClick={
                                    () => {
                                        onAddBikeOpen();
                                    }
                                }
                            >
                                Add
                            </Button>
                        )
                    }
                </HStack>
                <Divider />
                {
                    memberBikes && (
                        <UnorderedList pt={10} spacing={2}>
                            {
                                (memberBikes as Bike[]).map((bike) => (
                                    <HStack>
                                        <ListItem
                                            fontSize="xl"
                                        >
                                            {` ${bike.year} ${bike.make} ${bike.model}`}
                                        </ListItem>
                                        {
                                            props.admin && (
                                                <Button
                                                    textDecoration="underline"
                                                    color="red"
                                                    variant="ghost"
                                                    size="sm"
                                                    onClick={
                                                        () => {
                                                            setBikeToRemove(bike);
                                                            onRemoveBikeOpen();
                                                        }
                                                    }
                                                >
                                                    Remove
                                                </Button>
                                            )
                                        }
                                        {
                                            props.admin && (
                                                <Button
                                                    textDecoration="underline"
                                                    color="orange"
                                                    variant="ghost"
                                                    size="sm"
                                                    onClick={
                                                        () => {
                                                            setBikeToEdit(bike);
                                                            onEditBikeOpen();
                                                        }
                                                    }
                                                >
                                                    Edit
                                                </Button>
                                            )
                                        }
                                    </HStack>
                                ))
                            }
                        </UnorderedList>
                    )
                }
            </VStack>
            {
                isRemoveFamilyOpen && (
                    <DeleteAlert
                        isOpen={isRemoveFamilyOpen}
                        onClose={onRemoveFamilyClose}
                        // eslint-disable-next-line react/jsx-no-bind
                        removeMethod={removeFamilyMember}
                    />
                )
            }
            {
                isRemoveBikeOpen && (
                    <DeleteAlert
                        isOpen={isRemoveBikeOpen}
                        onClose={onRemoveBikeClose}
                        // eslint-disable-next-line react/jsx-no-bind
                        removeMethod={removeBike}
                    />
                )
            }
            {
                isEditBikeOpen && bikeToEdit && (
                    <EditBikesModal
                        isOpen={isEditBikeOpen}
                        onClose={onEditBikeClose}
                        bikeToEdit={bikeToEdit}
                        // eslint-disable-next-line react/jsx-no-bind
                        editBike={editBike}
                    />
                )
            }
            {
                isAddFamilyOpen && (
                    <AddFamilyModal
                        isOpen={isAddFamilyOpen}
                        onClose={onAddFamilyClose}
                        membershipAdmin={state.user}
                        token={state.token}
                    />
                )
            }
            {
                isAddBikeOpen && (
                    <AddBikeModal
                        isOpen={isAddBikeOpen}
                        onClose={onAddBikeClose}
                        // eslint-disable-next-line react/jsx-no-bind
                        addBike={addBike}
                    />
                )
            }
        </SimpleGrid>
    );
}
