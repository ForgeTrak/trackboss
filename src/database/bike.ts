import _ from 'lodash';
import { OkPacket, RowDataPacket } from 'mysql2';

import logger from '../logger';
import pool from './pool';
import { Bike, PostNewBikeRequest } from '../typedefs/bike';

export const INSERT_BIKE_SQL = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
export const GET_BIKE_SQL = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';

export async function insertBike(req: PostNewBikeRequest): Promise<number> {
    //const sql = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
    const values = [req.year, req.make, req.model, req.membershipId];

    let result;
    try {
        [result] = await pool.query<OkPacket>(INSERT_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error inserting bike: ${e}`);
        throw new Error('internal server error');
    }

    return result.insertId;
}

export async function getBike(id: number): Promise<Bike> {
    //const sql = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';
    const values = [id];

    let results;
    try {
        [results] = await pool.query<RowDataPacket[]>(GET_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error getting bike: ${e}`);
        throw new Error('internal server error');
    }

    if (_.isEmpty(results)) {
        throw new Error('not found');
    }

    return {
        bikeId: results[0].bike_id,
        year: results[0].year,
        make: results[0].make,
        model: results[0].model,
        membershipAdmin: results[0].membership_admin,
    };
}
