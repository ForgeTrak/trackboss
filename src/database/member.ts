import _ from 'lodash';
import { OkPacket, RowDataPacket } from 'mysql2';

import logger from '../logger';
import { getPool } from './pool';
import { GetMemberListFilters, Member, PatchMemberRequest, PostNewMemberRequest } from '../typedefs/member';
import { createCognitoUser } from '../util/cognito';

// Map the API values for the member types to the DB values
export const MEMBER_TYPE_MAP = new Map([
    ['admin', 'Admin'],
    ['membershipAdmin', 'Membership Admin'],
    ['member', 'Member'],
    ['paidLaborer', 'Paid Laborer'],
]);

export const GET_MEMBER_LIST_SQL = 'SELECT * FROM v_member';
export const GET_MEMBER_SQL = `${GET_MEMBER_LIST_SQL} WHERE member_id = ?`;
export const GET_MEMBER_UUID_SQL = `${GET_MEMBER_LIST_SQL} WHERE uuid = ?`;
export const INSERT_MEMBER_SQL = 'INSERT INTO member (membership_id, uuid, member_type_id, first_name, last_name, ' +
    'phone_number, occupation, email, birthdate, date_joined, last_modified_date, last_modified_by, active) ' +
    'VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURDATE(), ?, 1)';
export const PATCH_MEMBER_SQL = 'CALL sp_patch_member(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)';
export const GET_VALID_ACTORS_SQL = 'select member_id from pradb.member m where member_id=? or (m.member_type_id=2 ' +
    'and m.member_id=(select ms.membership_admin_id from pradb.member m left join pradb.membership ms on ' +
    'm.membership_id=ms.membership_id where member_id=?)) or m.member_type_id=1';

export async function insertMember(req: PostNewMemberRequest): Promise<number> {
    if (req.email) {
        logger.info(`Creating new user for email ${req.email} on membership ${req.membershipId}`);
        try {
            const isMembershipAdmin = (req.memberTypeId === 8);
            const uuid = await createCognitoUser(req.email, isMembershipAdmin);
            req.uuid = uuid;
            logger.info(`Successfully created ${req.email} in cognito as ${uuid}`);
        } catch (error: any) {
            logger.error(`Failure creating ${req.email} in cognito.  Continuing on trackboss database side`);
            logger.error(error);
        }
    }
    const values = [
        req.membershipId,
        req.uuid,
        req.memberTypeId,
        req.firstName,
        req.lastName,
        req.phoneNumber,
        req.occupation,
        req.email,
        req.birthdate,
        req.dateJoined,
        req.modifiedBy,
    ];

    let result;
    try {
        [result] = await getPool().query<OkPacket>(INSERT_MEMBER_SQL, values);
    } catch (e: any) {
        if ('errno' in e) {
            switch (e.errno) {
                case 1048: // non-null violation, missing a non-nullable column
                case 1452: // FK violation - referenced is missing
                    logger.error(`User error inserting member in DB: ${e}`);
                    throw new Error('user input error');
                default:
                    logger.error(`DB error inserting member: ${e}`);
                    throw new Error('internal server error');
            }
        } else {
            // this should not happen - errors from query should always have 'errno' field
            throw e;
        }
    }

    return result.insertId;
}

export async function getMemberList(filters: GetMemberListFilters): Promise<Member[]> {
    let sql;
    let values: any[] = [];
    if (!_.isEmpty(filters)) {
        let dynamicSql = ' WHERE ';
        let counter = 0;
        if (typeof filters.type !== 'undefined') {
            dynamicSql += 'member_type = ? AND ';
            // if type is not in the map, it won't hurt to throw it in
            // anyway (and this makes testing easier)
            values[counter++] = MEMBER_TYPE_MAP.get(filters.type) || filters.type;
        }
        if (typeof filters.membershipId !== 'undefined') {
            dynamicSql += 'membership_id = ? AND ';
            values[counter++] = filters.membershipId;
        }
        sql = GET_MEMBER_LIST_SQL + dynamicSql.slice(0, -4); // slice the trailing AND
    } else {
        sql = GET_MEMBER_LIST_SQL;
        values = [];
    }
    let results;
    try {
        [results] = await getPool().query<RowDataPacket[]>(sql, values);
    } catch (e) {
        logger.error(`DB error getting member list: ${e}`);
        throw new Error('internal server error');
    }

    return results.map((result) => ({
        memberId: result.member_id,
        membershipId: result.membership_id,
        firstName: result.first_name,
        lastName: result.last_name,
        membershipAdmin: result.membership_admin,
        membershipAdminId: result.membership_admin_id,
        isBoardMember: !!result.board_title_id,
        uuid: result.uuid,
        active: !!result.active[0],
        memberTypeId: result.member_type_id,
        memberType: result.member_type,
        membershipType: result.membership_type,
        phoneNumber: result.phone_number,
        occupation: result.occupation,
        email: result.email,
        birthdate: result.birthdate,
        dateJoined: result.date_joined,
        address: result.address,
        city: result.city,
        state: result.state,
        zip: result.zip,
        lastModifiedDate: result.last_modified_date,
        lastModifiedBy: result.last_modified_by,
    }));
}

export async function getMember(searchParam: string): Promise<Member> {
    const id = Number(searchParam);
    let sql;
    let values;
    let results;
    if (Number.isNaN(id)) {
        // uuid search
        values = [searchParam];
        sql = GET_MEMBER_UUID_SQL;
    } else {
        values = [id];
        sql = GET_MEMBER_SQL;
    }

    try {
        [results] = await getPool().query<RowDataPacket[]>(sql, values);
    } catch (e) {
        logger.error(`DB error getting member: ${e}`);
        throw new Error('internal server error');
    }

    if (_.isEmpty(results)) {
        throw new Error('not found');
    }

    return {
        memberId: results[0].member_id,
        membershipId: results[0].membership_id,
        firstName: results[0].first_name,
        lastName: results[0].last_name,
        membershipAdmin: results[0].membership_admin,
        membershipAdminId: results[0].membership_admin_id,
        isBoardMember: !!results[0].board_title_id,
        uuid: results[0].uuid,
        active: (results[0].active > 0),
        memberTypeId: results[0].member_type_id,
        memberType: results[0].member_type,
        membershipType: results[0].membership_type,
        phoneNumber: results[0].phone_number,
        occupation: results[0].occupation,
        email: results[0].email,
        birthdate: results[0].birthdate,
        dateJoined: results[0].date_joined,
        address: results[0].address,
        city: results[0].city,
        state: results[0].state,
        zip: results[0].zip,
        lastModifiedDate: results[0].last_modified_date,
        lastModifiedBy: results[0].last_modified_by,
    };
}

export async function getMemberByPhone(phone: string): Promise<Member> {
    let results;
    try {
        const sql = 'SELECT * FROM pradb.v_member where phone_number = ?';
        [results] = await getPool().query<RowDataPacket[]>(sql, [phone]);
    } catch (e) {
        logger.error(e);
        throw e;
    }
    return {
        memberId: results[0].member_id,
        membershipId: results[0].membership_id,
        firstName: results[0].first_name,
        lastName: results[0].last_name,
        membershipAdmin: results[0].membership_admin,
        membershipAdminId: results[0].membership_admin_id,
        isBoardMember: !!results[0].board_title_id,
        uuid: results[0].uuid,
        active: true,
        memberTypeId: results[0].member_type_id,
        memberType: results[0].member_type,
        membershipType: results[0].membership_type,
        phoneNumber: results[0].phone_number,
        occupation: results[0].occupation,
        email: results[0].email,
        birthdate: results[0].birthdate,
        dateJoined: results[0].date_joined,
        address: results[0].address,
        city: results[0].city,
        state: results[0].state,
        zip: results[0].zip,
        lastModifiedDate: results[0].last_modified_date,
        lastModifiedBy: results[0].last_modified_by,
    };
}

export async function getMemberByEmail(email: string): Promise<any> {
    let results;
    try {
        const sql = `SELECT * FROM pradb.v_member where email like '%${email}%'`;
        [results] = await getPool().query<RowDataPacket[]>(sql, [email]);
    } catch (e) {
        logger.error(e);
        throw e;
    }
    let memberWithEmail = {};
    if (results?.length > 0) {
        memberWithEmail = {
            memberId: results[0].member_id,
            membershipId: results[0].membership_id,
            firstName: results[0].first_name,
            lastName: results[0].last_name,
            membershipAdmin: results[0].membership_admin,
            membershipAdminId: results[0].membership_admin_id,
            uuid: results[0].uuid,
            active: true,
            memberTypeId: results[0].member_type_id,
            memberType: results[0].member_type,
            membershipType: results[0].membership_type,
            phoneNumber: results[0].phone_number,
            occupation: results[0].occupation,
            email: results[0].email,
            birthdate: results[0].birthdate,
            dateJoined: results[0].date_joined,
            address: results[0].address,
            city: results[0].city,
            state: results[0].state,
            zip: results[0].zip,
            lastModifiedDate: results[0].last_modified_date,
            lastModifiedBy: results[0].last_modified_by,
        };
    }
    return memberWithEmail;
}

export async function patchMember(id: string, req: PatchMemberRequest): Promise<void> {
    if (_.isEmpty(req)) { // empty request
        throw new Error('user input error');
    }
    const values = [
        id,
        req.membershipId,
        req.uuid,
        req.active,
        req.memberTypeId,
        req.firstName,
        req.lastName,
        req.phoneNumber,
        req.occupation,
        req.email,
        req.birthdate,
        req.dateJoined,
        req.modifiedBy,
    ];
    let result;
    try {
        [result] = await getPool().query<OkPacket>(PATCH_MEMBER_SQL, values);
    } catch (e: any) {
        if ('errno' in e) {
            switch (e.errno) {
                case 1451: // FK violation - referenced somewhere else
                case 1452: // FK violation - referenced is missing
                    logger.error(`User error patching member in DB: ${e}`);
                    throw new Error('user input error');
                default:
                    logger.error(`DB error patching member: ${e}`);
                    throw new Error('internal server error');
            }
        } else {
            // this should not happen - errors from query should always have 'errno' field
            throw e;
        }
    }

    if (result.affectedRows < 1) {
        throw new Error('not found');
    }
}

export async function getValidActors(member: number): Promise<number[]> {
    const values = [member, member];
    let results;
    try {
        [results] = await getPool().query<RowDataPacket[]>(GET_VALID_ACTORS_SQL, values);
    } catch (e) {
        logger.error(`DB error getting valid actors: ${e}`);
        throw new Error('internal server error');
    }
    if (_.isEmpty(results)) {
        throw new Error('not found');
    }
    return _.map(results, (result) => result.member_id);
}

export async function deleteFamilyMember(memberId: number): Promise<number> {
    // remove the member but only if they are a sub member. Magic numberism here, fix it later.
    const sql =
        'delete from member where member_id = ? and member_type_id = 9';
    const values = [memberId];

    let result;
    try {
        [result] = await getPool().query<OkPacket>(sql, values);
    } catch (e) {
        logger.error(`DB error deleting event: ${e}`);
        throw new Error('internal server error');
    }
    return result.affectedRows;
}
