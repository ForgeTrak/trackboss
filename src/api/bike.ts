import { Request, Response, Router } from 'express';
import { deleteBike, getBike, getBikeList, insertBike, patchBike } from '../database/bike';
import {
    Bike,
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeResponse,
    PostNewBikeResponse,
} from '../typedefs/bike';
import { checkHeader, verify } from '../util/auth';

const bike = Router();

bike.post('/new', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PostNewBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token, 'Membership Admin');
            const insertId = await insertBike(req.body);
            response = await getBike(insertId);
            res.status(201);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/list', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeListResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const id = req.query.membershipID;
            let filterMembership: number | undefined;
            if (typeof id === 'undefined') {
                filterMembership = undefined;
            } else {
                filterMembership = Number(id);
                if (Number.isNaN(filterMembership)) {
                    throw new Error('user input error');
                }
            }
            const bikeList: Bike[] = await getBikeList(filterMembership);
            res.status(200);
            response = bikeList;
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const { bikeID } = req.params;
            response = await getBike(Number(bikeID));
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.patch('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PatchBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await patchBike(bikeIdNum, req.body);
            response = await getBike(bikeIdNum);
            res.status(200);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.delete('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: DeleteBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await deleteBike(bikeIdNum);
            response = { bikeId: bikeIdNum };
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

export default bike;
