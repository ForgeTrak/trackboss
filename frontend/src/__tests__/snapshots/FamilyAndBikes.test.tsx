import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import FamilyAndBikes from '../../components/FamilyAndBikes';
import { Member } from '../../../../src/typedefs/member';
import { Bike } from '../../../../src/typedefs/bike';

const memberFamily: Member[] = [
    {
        memberId: 1,
        membershipAdmin: 'true',
        active: true,
        memberType: 'admin',
        firstName: 'John',
        lastName: 'Smith',
        phoneNumber: '1',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: 'August 12, 2006',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
    {
        memberId: 2,
        membershipAdmin: 'true',
        active: true,
        memberType: 'admin',
        firstName: 'Jane',
        lastName: 'Smith',
        phoneNumber: '2',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: 'August 12, 2006',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
];

const memberBikes: Bike[] = [
    {
        bikeId: 1,
        year: '2012',
        make: 'honda',
        model: 'rust-bucket',
        membershipAdmin: 'string',
    },
    {
        bikeId: 2,
        year: '2022',
        make: 'yamaha',
        model: '',
        membershipAdmin: 'string',
    },
];

describe('family and bikes', () => {
    it('renders all props correctly as admin', () => {
        const familyAndBikes = render(
            <FamilyAndBikes memberBikes={memberBikes} memberFamily={memberFamily} admin />,
        );
        expect(familyAndBikes).toMatchSnapshot();
    });

    it('renders all props correctly as not admin', () => {
        const familyAndBikes = render(
            <FamilyAndBikes memberBikes={memberBikes} memberFamily={memberFamily} admin={false} />,
        );
        expect(familyAndBikes).toMatchSnapshot();
    });
});
