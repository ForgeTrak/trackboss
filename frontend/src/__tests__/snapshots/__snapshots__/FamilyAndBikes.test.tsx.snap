// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`family and bikes renders all props correctly as admin 1`] = `[Function]`;

exports[`family and bikes renders all props correctly as not admin 1`] = `[Function]`;
